packages/app/src/elements/audit.ts

```diff
@@ -0,0 +1,311 @@
+                                ></pl-vault-item-list-item>
+                            </div>
+                        `
+                    )}
+                </pl-list>
+                <pl-button
+                    class="slim margined transparent"
+                    @click=${() => this.go("items", routeParams)}
+                    ?hidden=${listItems.length < 6}
+                >
+                    <div>${$l("Show All")}</div>
+                    <pl-icon icon="arrow-right" class="small left-margined"></pl-icon>
+                </pl-button>
+            </section>
+        `;
+    }
+}
+
+async function sha1(password: string) {
+    const hashedPasswordData = await getCryptoProvider().hash(
+        stringToBytes(password),
+        new HashParams({ algorithm: "SHA-1" })
+    );
+    const hashedPassword = bytesToHex(hashedPasswordData);
+    return hashedPassword;
+}
+
+async function isPasswordWeak(password: string) {
+    const { score } = await passwordStrength(password);
+
+    return score < 2;
+}
+
+async function hasPasswordBeenCompromised(passwordHash: string) {
+    const hashPrefix = passwordHash.slice(0, 5);
+
+    const response = await fetch(`https://api.pwnedpasswords.com/range/${hashPrefix}`);
+
+    const result = await response.text();
+
+    const matchingHashSuffixes = result.split("\r\n");
+
+    for (const matchingHashSuffix of matchingHashSuffixes) {
+        const fullLowercaseHash = `${hashPrefix}${matchingHashSuffix.toLowerCase().split(":")[0]}`;
+
+        if (fullLowercaseHash === passwordHash) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+export async function auditVaults(
+    vaults: Vault[],
+    {
+        updateOnlyItemWithId,
+        updateOnlyIfOutdated,
+    }: { updateOnlyItemWithId?: string; updateOnlyIfOutdated?: boolean } = {}
+) {
+    const reusedPasswords: ListItem[] = [];
+    const weakPasswords: ListItem[] = [];
+    const compromisedPasswords: ListItem[] = [];
+
+    const firstMatchPerPasswordHash = new Map<
+        string,
+        { item: VaultItem; vault: Vault; passwordField: { field: Field; fieldIndex: number } }
+    >();
+    const reusedPasswordItemIds: Set<string> = new Set();
+    const weakPasswordItemIds: Set<string> = new Set();
+    const compromisedPasswordItemIds: Set<string> = new Set();
+
+    const oneWeekAgo = sub(new Date(), { weeks: 1 });
+
+    // TODO: Remove this
+    console.log(`Running audit! ${JSON.stringify({ updateOnlyItemWithId, updateOnlyIfOutdated })}`);
+
+    for (const vault of vaults) {
+        for (const item of vault.items) {
+            if (updateOnlyItemWithId) {
+                if (item.id !== updateOnlyItemWithId) {
+                    // TODO: Remove this
+                    console.log("Skipped, not the matching item.");
+                    continue;
+                }
+            }
+
+            if (updateOnlyIfOutdated && item.lastAudited && item.lastAudited >= oneWeekAgo) {
```

Comment:
I chose to not run _any_ audit if it's been audited recently, because we already run an audit if the item changes, so it seems very unnecessary.

Reply:
The idea was to also call this method even when a single item is updated. This may seem a bit inefficient but it shouldn't take up too many cycles since we'll be skipping most items anyway.

Reply:
Never mind... I misread your comment.

Reply:
Not running the audit when an item is updated makes it seem like a now "good" password still looks bad (with failed audits), that's why I wanted to run it there. It'll also _only_ run for that item (no matter when the last time it ran), not for everything.

Am I missing something on why that would be bad?

Reply:
The audit should definitely be run whenever an item is updated, the question is only how that action should be triggered. Since after saving an item the condition `item.updated  > item.lastAudited` is always true, the execution path I described would cause the audit to re-run automatically on the next "tick" (and after whatever timeout we set for the debounce). I hadn't considered the use case of "fixing" an audit warning, where you'd want to see the warning disappear immediately, but this could easily be solved in a few different ways:

- Use [throttle](https://github.com/padloc/padloc/blob/v4/packages/core/src/util.ts#L123) instead of [debounce](https://github.com/padloc/padloc/blob/v4/packages/core/src/util.ts#L101) (other than debounce, throttle executes the first invocation immediately).
- Delete the `auditResults` and `lastAudited` properties from the item when updating fields (this would make the warnings disappear immediately. Any new warnings would appear after a short delay).
- Call `runAudits` explicitly at the end of [`ItemView::save()`](https://github.com/padloc/padloc/blob/v4/packages/app/src/elements/item-view.ts#L493) (No `updateOnlyIfOutdated` parameter needed though since, like I said `item.updated  > item.lastAudited` will always be true after an item was saved).
- A combination of the above.

So there is nothing that speaks against explicitly triggering a re-audit (other than that it may be redundant), I just don't want to have to rely on "remembering" to trigger a re-audit whenever we make changes to an item, since there may be places other than the item view where that may happen.

Reply:
Gotcha, I'll try to implement it a bit differently, then (removing the warnings would help even if any new warning eventually shows up again).